src/executables/sysrepo-plugind.c

```diff
@@ -518,6 +528,34 @@ main(int argc, char **argv)
         goto cleanup;
     }
 
+    /* open and lock pid file before plugins are loaded, but don't update it yet 
+     * plugins may fail, and do some initialization that must finish before we are
+     * considered up and running. No need to create a pid file if we're in debug
+     * mode (we're not a daemon)
+     */
+
+    if (debug && pidfile_specifid) {
+        SRP_LOG_WRN("Ignoring pidfile in debug mode.");
+    }
+
+    if (!debug) {
+        /* make sure we are the only instance - lock the PID file and write the PID */
+        pidfd = open(pidfile, O_RDWR | O_CREAT, 0640);
+        if (pidfd < 0) {
+            error_print(0, "Unable to open the PID file \"%s\" (%s).", pidfile, strerror(errno));
+            goto cleanup;
+        }
+
+        if (lockf(pidfd, F_TLOCK, 0) < 0) {
+            if ((errno == EACCES) || (errno == EAGAIN)) {
+                error_print(0, "Another instance of the sysrepo-plugind is running.");
+            } else {
+                error_print(0, "Unable to lock the PID file \"%s\" (%s).", pidfile, strerror(errno));
+            }
+            goto cleanup;
```

Comment:
This change will effectively make it impossible to run `sysrepo-plugind` as non-`root` by default. Since `/var/run` is not writable by all, unless people take an explicit action to build with an option to put the PID file into a writable location, `sysrepo-plugind` with no arguments will fail to start on a developer's box.

Instead of a build-time option for a default PID file location, and then a runtime option for overriding that, why not just do nothing by default, *and* allow the user to pass a flag like `--pid-file=/var/run/sysrepo-plugind.pid` which will activate this mechanism?

Reply:
Great idea, thanks!